File: src/ISA/ISA/JsonTypes/Component.fs
namespace ISA

open ISA.Aux

type Component = 
    {
        // TODO: Maybe remove as field and add as member?
        ComponentName : string option
        /// This can be the main column value of the component column. (e.g. "SCIEX instrument model" as `OntologyAnnotation`; 14;..)
        ComponentValue : Value option
        /// This can be the unit describing a non `OntologyAnnotation` value in `ComponentValue`. (e.g. "degree celcius")
        ComponentUnit : OntologyAnnotation option
        /// This can be the component column header (e.g. "instrument model")
        ComponentType : OntologyAnnotation option
    }

    static member make name value unit componentType =
        {
            ComponentName = name
            ComponentValue = value
            ComponentUnit = unit
            ComponentType = componentType
        }

    static member create (?Name,?Value,?Unit,?ComponentType) : Component =
        Component.make Name Value Unit ComponentType

    static member empty =
        Component.create()
    
    /// This function creates a string containing full isa term triplet information about the component
    ///
    /// Components do not have enough fields in ISA-JSON to include all existing ontology term information. 
    /// This function allows us, to add the same information as `Parameter`, `Characteristics`.., to `Component`. 
    /// Without this string composition we loose the ontology information for the header value.
    static member composeName (value : Value Option) (unit : OntologyAnnotation option) = 
        match value,unit with
        | Some (Value.Ontology oa), _ ->
            $"{oa.NameText} ({oa.TermAccessionShort})"
        | Some v, None ->
            $"{v.AsString}"
        | Some v, Some u ->
            $"{v.AsString} {u.NameText} ({u.TermAccessionShort})"
        | None, _ -> ""

    /// This function parses the given Component header string format into the ISA-JSON Component type
    ///
    /// Components do not have enough fields in ISA-JSON to include all existing ontology term information. 
    /// This function allows us, to add the same information as `Parameter`, `Characteristics`.., to `Component`. 
    /// Without this string composition we loose the ontology information for the header value.
    static member decomposeName (name : string) = 
        let pattern = """(?<value>[^\(]+) \((?<ontology>[^(]*:[^)]*)\)"""
        let unitPattern = """(?<value>[\d\.]+) (?<unit>.+) \((?<ontology>[^(]*:[^)]*)\)"""

        let r = System.Text.RegularExpressions.Regex.Match(name,pattern)
        let unitr = System.Text.RegularExpressions.Regex.Match(name,unitPattern)

        if unitr.Success then
            let oa = (unitr.Groups.Item "ontology").Value   |> OntologyAnnotation.fromTermAccession 
            let v =  (unitr.Groups.Item "value").Value      |> Value.fromString
            let u =  (unitr.Groups.Item "unit").Value
            v, Some {oa with Name = (Some (AnnotationValue.Text u))}
        elif r.Success then
            let oa = (r.Groups.Item "ontology").Value   |> OntologyAnnotation.fromTermAccession 
            let v =  (r.Groups.Item "value").Value      |> Value.fromString
            Value.Ontology {oa with Name = (Some (AnnotationValue.Text v.AsString))}, None
        else 
            Value.Name (name), None       

    /// Create a ISAJson Component from ISATab string entries
    static member fromString (?name: string, ?term:string, ?source:string, ?accession:string, ?comments : Comment list) = 
        let cType = OntologyAnnotation.fromString (?term = term, ?tsr=source, ?tan=accession, ?comments = comments) |> Option.fromValueWithDefault OntologyAnnotation.empty
        match name with
        | Some n -> 
            let v,u = Component.decomposeName n
            Component.make (name) (Option.fromValueWithDefault (Value.Name "") v) u cType
        | None ->
            Component.make None None None cType
        
    static member fromOptions (value: Value option) (unit: OntologyAnnotation Option) (header:OntologyAnnotation option) = 
        let name = Component.composeName value unit |> Option.fromValueWithDefault ""
        Component.make name value unit header

    /// Get ISATab string entries from an ISAJson Component object
    static member toString (c : Component) =
        let oa = c.ComponentType |> Option.map OntologyAnnotation.toString |> Option.defaultValue {|TermName = ""; TermAccessionNumber = ""; TermSourceREF = ""|}
        c.ComponentName |> Option.defaultValue "", oa

    member this.NameText =
        this.ComponentType
        |> Option.map (fun c -> c.NameText)
        |> Option.defaultValue ""

    /// Returns the ontology of the category of the Value as string
    member this.UnitText = 
        this.ComponentUnit
        |> Option.map (fun c -> c.NameText)
        |> Option.defaultValue ""

    member this.ValueText = 
        this.ComponentValue
        |> Option.map (fun c -> c.AsString)
        |> Option.defaultValue ""

    member this.ValueWithUnitText =
        let unit = 
            this.ComponentUnit |> Option.map (fun oa -> oa.NameText)
        let v = this.ValueText
        match unit with
        | Some u    -> sprintf "%s %s" v u
        | None      -> v

    member this.MapCategory(f : OntologyAnnotation -> OntologyAnnotation) =
        {this with ComponentType = Option.map f this.ComponentType}

    member this.SetCategory(c : OntologyAnnotation) =
        {this with ComponentType = Some c}

File: js/ISA/ISA/JsonTypes/Component.js
import { Record } from "../../../fable_modules/fable-library.4.1.4/Types.js";
import { record_type, option_type, string_type } from "../../../fable_modules/fable-library.4.1.4/Reflection.js";
import { Value as Value_1, Value_fromString_Z721C83C5, Value__get_Text, Value_$reflection } from "./Value.js";
import { OntologyAnnotation_toString_5E3DAF0D, OntologyAnnotation_get_empty, OntologyAnnotation_fromString_2EB0E147, OntologyAnnotation, OntologyAnnotation_fromTermAnnotation_Z721C83C5, OntologyAnnotation__get_TermAccessionShort, OntologyAnnotation__get_NameText, OntologyAnnotation_$reflection } from "./OntologyAnnotation.js";
import { create, match } from "../../../fable_modules/fable-library.4.1.4/RegExp.js";
import { AnnotationValue } from "./AnnotationValue.js";
import { map, defaultArg, unwrap } from "../../../fable_modules/fable-library.4.1.4/Option.js";
import { fromValueWithDefault } from "../OptionExtensions.js";
import { printf, toText } from "../../../fable_modules/fable-library.4.1.4/String.js";

export class Component extends Record {
    constructor(ComponentName, ComponentValue, ComponentUnit, ComponentType) {
        super();
        this.ComponentName = ComponentName;
        this.ComponentValue = ComponentValue;
        this.ComponentUnit = ComponentUnit;
        this.ComponentType = ComponentType;
    }
}

export function Component_$reflection() {
    return record_type("ARCtrl.ISA.Component", [], Component, () => [["ComponentName", option_type(string_type)], ["ComponentValue", option_type(Value_$reflection())], ["ComponentUnit", option_type(OntologyAnnotation_$reflection())], ["ComponentType", option_type(OntologyAnnotation_$reflection())]]);
}

export function Component_make(name, value, unit, componentType) {
    return new Component(name, value, unit, componentType);
}

export function Component_create_61502994(Name, Value, Unit, ComponentType) {
    return Component_make(Name, Value, Unit, ComponentType);
}

export function Component_get_empty() {
    return Component_create_61502994();
}

/**
 * This function creates a string containing full isa term triplet information about the component
 * 
 * Components do not have enough fields in ISA-JSON to include all existing ontology term information.
 * This function allows us, to add the same information as `Parameter`, `Characteristics`.., to `Component`.
 * Without this string composition we loose the ontology information for the header value.
 */
export function Component_composeName(value, unit) {
    if (value == null) {
        return "";
    }
    else if (value.tag === 0) {
        const oa = value.fields[0];
        return `${OntologyAnnotation__get_NameText(oa)} (${OntologyAnnotation__get_TermAccessionShort(oa)})`;
    }
    else if (unit != null) {
        const u = unit;
        const v_1 = value;
        return `${Value__get_Text(v_1)} ${OntologyAnnotation__get_NameText(u)} (${OntologyAnnotation__get_TermAccessionShort(u)})`;
    }
    else {
        const v = value;
        return `${Value__get_Text(v)}`;
    }
}

/**
 * This function parses the given Component header string format into the ISA-JSON Component type
 * 
 * Components do not have enough fields in ISA-JSON to include all existing ontology term information.
 * This function allows us, to add the same information as `Parameter`, `Characteristics`.., to `Component`.
 * Without this string composition we loose the ontology information for the header value.
 */
export function Component_decomposeName_Z721C83C5(name) {
    const r = match(create("(?<value>[^\\(]+) \\((?<ontology>[^(]*:[^)]*)\\)"), name);
    const unitr = match(create("(?<value>[\\d\\.]+) (?<unit>.+) \\((?<ontology>[^(]*:[^)]*)\\)"), name);
    if (unitr != null) {
        const oa = OntologyAnnotation_fromTermAnnotation_Z721C83C5((unitr.groups && unitr.groups.ontology) || "");
        return [Value_fromString_Z721C83C5((unitr.groups && unitr.groups.value) || ""), new OntologyAnnotation(oa.ID, new AnnotationValue(0, [(unitr.groups && unitr.groups.unit) || ""]), oa.TermSourceREF, oa.LocalID, oa.TermAccessionNumber, oa.Comments)];
    }
    else if (r != null) {
        const oa_1 = OntologyAnnotation_fromTermAnnotation_Z721C83C5((r.groups && r.groups.ontology) || "");
        return [new Value_1(0, [new OntologyAnnotation(oa_1.ID, new AnnotationValue(0, [Value__get_Text(Value_fromString_Z721C83C5((r.groups && r.groups.value) || ""))]), oa_1.TermSourceREF, oa_1.LocalID, oa_1.TermAccessionNumber, oa_1.Comments)]), void 0];
    }
    else {
        return [new Value_1(3, [name]), void 0];
    }
}

/**
 * Create a ISAJson Component from ISATab string entries
 */
export function Component_fromString_Z61E08C1(name, term, source, accession, comments) {
    let cType;
    const v = OntologyAnnotation_fromString_2EB0E147(unwrap(term), unwrap(source), unwrap(accession), unwrap(comments));
    cType = fromValueWithDefault(OntologyAnnotation_get_empty(), v);
    if (name == null) {
        return Component_make(void 0, void 0, void 0, cType);
    }
    else {
        const patternInput = Component_decomposeName_Z721C83C5(name);
        return Component_make(name, fromValueWithDefault(new Value_1(3, [""]), patternInput[0]), patternInput[1], cType);
    }
}

export function Component_fromOptions(value, unit, header) {
    return Component_make(fromValueWithDefault("", Component_composeName(value, unit)), value, unit, header);
}

/**
 * Get ISATab string entries from an ISAJson Component object
 */
export function Component_toString_Z609B8895(c) {
    let oa;
    const value = {
        TermAccessionNumber: "",
        TermName: "",
        TermSourceREF: "",
    };
    oa = defaultArg(map(OntologyAnnotation_toString_5E3DAF0D, c.ComponentType), value);
    return [defaultArg(c.ComponentName, ""), oa];
}

export function Component__get_NameText(this$) {
    return defaultArg(map(OntologyAnnotation__get_NameText, this$.ComponentType), "");
}

/**
 * Returns the ontology of the category of the Value as string
 */
export function Component__get_UnitText(this$) {
    return defaultArg(map(OntologyAnnotation__get_NameText, this$.ComponentUnit), "");
}

export function Component__get_ValueText(this$) {
    return defaultArg(map(Value__get_Text, this$.ComponentValue), "");
}

export function Component__get_ValueWithUnitText(this$) {
    const unit = map(OntologyAnnotation__get_NameText, this$.ComponentUnit);
    const v = Component__get_ValueText(this$);
    if (unit == null) {
        return v;
    }
    else {
        const u = unit;
        return toText(printf("%s %s"))(v)(u);
    }
}

export function Component__MapCategory_Z69DD836A(this$, f) {
    return new Component(this$.ComponentName, this$.ComponentValue, this$.ComponentUnit, map(f, this$.ComponentType));
}

export function Component__SetCategory_Z4C0FE73C(this$, c) {
    return new Component(this$.ComponentName, this$.ComponentValue, this$.ComponentUnit, c);
}

